olp-cpp-sdk-authentication/src/AuthenticationClient.cpp

```diff
@@ -775,6 +820,71 @@ client::CancellationToken AuthenticationClient::Impl::SignOut(
   });
 }
 
+client::CancellationToken AuthenticationClient::Impl::IntrospectApp(
+    std::string access_token, IntrospectAppCallback callback) {
+  auto introspect_app_task = [=](client::CancellationContext context)
+      -> client::ApiResponse<IntrospectAppResult, client::ApiError> {
+    if (!settings_.network_request_handler) {
+      AuthenticationError result({static_cast<int>(http::ErrorCode::IO_ERROR),
+                                  "Cannot introspectt app while offline"});
+      callback(std::move(result));
+      return client::ApiError();
+    }
+    std::weak_ptr<http::Network> weak_network(
+        settings_.network_request_handler);
+    std::string url = settings_.token_endpoint_url + kIntrospectAppEndpoint;
+    http::NetworkSettings network_settings;
+    if (settings_.network_proxy_settings) {
+      network_settings.WithProxySettings(
+          settings_.network_proxy_settings.get());
+    }
+    auto request =
+        http::NetworkRequest(url)
+            .WithVerb(http::NetworkRequest::HttpVerb::GET)
+            .WithHeader(http::kAuthorizationHeader,
+                        generateBearerHeader(access_token))
+            .WithHeader(http::kUserAgentHeader, http::kOlpSdkUserAgent)
+            .WithSettings(std::move(network_settings));
+
+    auto payload = std::make_shared<std::stringstream>();
+    auto send_outcome = settings_.network_request_handler->Send(
+        request, payload,
+        [callback, payload](const http::NetworkResponse& network_response) {
+          auto response_status = network_response.GetStatus();
+          auto error_msg = network_response.GetError();
+
+          if (response_status < 0) {
+            // Network error response not available
+            AuthenticationError result(response_status, error_msg);
+            callback(std::move(result));
+            return;
+          }
+
+          auto document = std::make_shared<rapidjson::Document>();
+          rapidjson::IStreamWrapper stream(*payload);
+          document->ParseStream(stream);
+
+          std::shared_ptr<IntrospectAppResultImpl> resp_impl =
+              std::make_shared<IntrospectAppResultImpl>(response_status,
+                                                        error_msg, document);
+          IntrospectAppResult response(resp_impl);
+          callback(std::move(response));
+        });
+
+    if (!send_outcome.IsSuccessful()) {
+      std::string error_message =
+          ErrorCodeToString(send_outcome.GetErrorCode());
+      AuthenticationError result({static_cast<int>(send_outcome.GetErrorCode()),
+                                  std::move(error_message)});
+      callback(std::move(result));
+    }
+    return client::ApiError();
+  };
+
+  return AddTask(settings_.task_scheduler, pending_requests_,
+                 std::move(introspect_app_task), nullptr);
```

Comment:
Instead of using the callback above you can cluster all that together and make use of the AddTask callback.

```cpp
using IntrospectAppResponse = client::ApiResponse<IntrospectAppResult, client::ApiError>;

auto callback = [](IntrospectAppResponse response){
   if (!Response.IsSuccessfull()) {
      // Convert from ApiError to AutheticationError
     callback(std::move(error));
   }

   callback(response.MoveResult());
};

AddTask(settings_.task_scheduler, pending_requests_,
                 std::move(introspect_app_task), callback);
```